### PR TITLE
[FX] Normalize torch.nn.functional calls and standard Module calls

### PR DESCRIPTION
**Pull request by jamesr66a**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#51816 [FX] Normalize torch.nn.functional calls and standard Module calls**

Add a Transfomer that:

* Normalizes calls to `torch.nn.functional*` functions s.t. we use the function signature to move all args to kwargs in the correct positional order
* Does the same for stdlib `nn.Module` instances

Differential Revision: [D26290764](https://our.internmc.facebook.com/intern/diff/D26290764)

### CODE REVIEW

File: test/test_fx_experimental.py [hunk 1]
```diff
@@ -749,6 +751,97 @@ def test_subgraph_trivial_resnet(self):
         module_with_submodules = split_module(traced, m, lambda node: 0)
         module_with_submodules(a)
 
+    @skipIfNoTorchVision
+    def test_normalize_args(self):
+        m = resnet18()
+
+        class FunctionalTracer(torch.fx.Tracer):
+            def is_leaf_module(self, m: torch.nn.Module, module_qualified_name : str) -> bool:
```
Comment (Chillee) [inline test/test_fx_experimental.py:759]: So you want to decompose all the modules except for BatchNorm?

File: test/test_fx_experimental.py [hunk 2]
```diff
@@ -749,6 +751,97 @@ def test_subgraph_trivial_resnet(self):
...
+                leaves = set([torch.nn.BatchNorm2d])
```
Comment (Chillee) [inline test/test_fx_experimental.py:760]: So you want to decompose all the modules except for BatchNorm?

File: test/test_fx_experimental.py [hunk 3]
```diff
@@ -749,6 +751,97 @@ def test_subgraph_trivial_resnet(self):
...
+                if submod_class == nn_class:
+                    self.assertEqual(len(node.args), 0)
+
+    def test_normalize_modules_exhaustive(self):
+        """
+        Exhaustively test `NormalizeArgs` on all standard
+        torch.nn Module classes
+        """
+        for test_params in module_tests + new_module_tests:
+            if 'constructor' not in test_params:
```
Comment (Chillee) [inline test/test_fx_experimental.py:789]: I assume these are special case things, might be worth adding a brief comment to explain why.

File: torch/fx/experimental/normalize.py [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+                 normalize_modules : bool = True):
+        super().__init__(module)
+        self.normalize_functionals = normalize_functionals
+        self.normalize_modules = normalize_modules
+
+    def call_function(self, target : Target, args : Tuple[Argument, ...], kwargs : Dict[str, Any]):
+        if self.normalize_functionals and target.__module__ == 'torch.nn.functional':
+            target_for_analysis = target
+            if target in boolean_dispatched:
+                # HACK: `boolean_dispatch` as used in `torch.nn.functional` makes it so that we have
```
Comment (Chillee) [inline torch/fx/experimental/normalize.py:28]: 🤔 Why does this exist lol.

File: torch/fx/experimental/normalize.py [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+        return super().call_function(target, args, kwargs)
+
+    def call_module(self, target : Target, args : Tuple[Argument, ...], kwargs : Dict[str, Any]):
+        assert isinstance(target, str)
+        submod = self.fetch_attr(target)
+        if self.normalize_modules and hasattr(submod.__class__, '__name__'):
+            classname = submod.__class__.__name__
+            if getattr(torch.nn, classname, None) == submod.__class__:
+                new_kwargs = self._args_kwargs_to_normalized_kwargs(submod.forward, args, kwargs)
+                if new_kwargs:
```
Comment (Chillee) [inline torch/fx/experimental/normalize.py:55]: Should this be an `is None`?

File: torch/fx/experimental/normalize.py [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+                dispatched = boolean_dispatched[target]
+                if_true, if_false = dispatched['if_true'], dispatched['if_false']
+                if inspect.signature(if_true).parameters != inspect.signature(if_false).parameters:
+                    return super().call_function(target, args, kwargs)
+                target_for_analysis = if_true
+
+            new_kwargs = self._args_kwargs_to_normalized_kwargs(target_for_analysis, args, kwargs)
+
+            if new_kwargs:
+                # FIXME: `target(**kwargs)` doesn't keep things specified as kwargs
```
Comment (Chillee) [inline torch/fx/experimental/normalize.py:42]: Maybe also specify that this is a hack. I thought it was some sort of limitation of the current code.

**Review by Chillee — APPROVED** [review]
LGTM

File: test/test_fx_experimental.py (hunk 2 above)
Comment (jamesr66a) [reply test/test_fx_experimental.py:760]: For testing purposes I want to trace everything down to the functional level, but BatchNorm2d is currently not traceable due to control flow content (see https://fb.quip.com/EgT5A3DFrpaJ#IbQACAPoQ0n)

File: torch/fx/experimental/normalize.py (hunk 4 above)
Comment (jamesr66a) [reply torch/fx/experimental/normalize.py:28]: Because TorchScript (of course lol) https://github.com/pytorch/pytorch/pull/14425

File: torch/fx/experimental/normalize.py (hunk 6 above)
Comment (jamesr66a) [reply torch/fx/experimental/normalize.py:42]: This seems to be a limitation of `__torch_function__`. It seems to normalize args differently (i.e. it looks up schema and puts positional args in `args` and keyword args in `kwargs`). Not 100% sure though but I know `__torch_function__` does pull schemas

File: torch/fx/experimental/normalize.py (hunk 5 above)
Comment (jamesr66a) [reply torch/fx/experimental/normalize.py:55]: No, `_args_kwargs_to_normalized_kwargs` returns `None` if this schema is not supported for normalization. We want to emit a `call_module` with `new_kwargs` if that is _not_ None

File: test/test_fx_experimental.py (hunk 3 above)
Comment (jamesr66a) [reply test/test_fx_experimental.py:789]: Honestly not sure why this is the way it is. This is copypasta from `torch_nn`: https://github.com/pytorch/pytorch/blob/b2d8f0a431f8ce2c606ed25ac4d28287489c71bc/test/test_nn.py#L10063

What i'm trying to do here is just get instances + example inputs for all the stdlib `nn.Module` instances

File: test/test_fx_experimental.py (hunk 2 above)
Comment (Chillee) [reply test/test_fx_experimental.py:760]: Might be worth adding a note.
